src/Bash.cpp

```diff
@@ -0,0 +1,297 @@
+        ICommand* command = newCommands.front();
+        _commands.emplace(command);
+        newCommands.pop();
+    }
+
+    // Copy the vector of path directories
+    _pathDirectories = other._pathDirectories;
+
+    // Copy the pipe file descriptors
+    _prevPipe[0] = other._prevPipe[0];
+    _prevPipe[1] = other._prevPipe[1];
+
+    // Copy the input/output file descriptors
+    _currPipe[0] = other._currPipe[0];
+    _currPipe[1] = other._currPipe[1];
+}
+
+/**
+ * Move constructor for the Bash class.
+ * Moves the queue of ICommand pointers, the vector of path directories,
+ * and the pipe file descriptors from the source object to the destination object.
+ * @param other The Bash object to be moved.
+ */
+Bash::Bash(Bash&& other) noexcept {
+    // Move the queue of ICommand pointers
+    _commands = std::move(other._commands);
+
+    // Move the vector of path directories
+    _pathDirectories = std::move(other._pathDirectories);
+
+    // Move the pipe file descriptors
+    _prevPipe[0] = other._prevPipe[0];
+    _prevPipe[1] = other._prevPipe[1];
+    other._prevPipe[0] = 0;
+    other._prevPipe[1] = 0;
+
+    // Move the input/output file descriptors
+    _currPipe[0] = other._currPipe[0];
+    _currPipe[1] = other._currPipe[1];
+    other._currPipe[0] = 0;
+    other._currPipe[1] = 0;
+}
+
+/**
+ * Copy assignment operator for the Bash class.
+ * Performs a deep copy of the queue of ICommand pointers, the vector of path directories,
+ * and the pipe file descriptors from the source object to the destination object.
+ * @param other The Bash object to be copied.
+ * @return The reference to the destination object.
+ */
+Bash& Bash::operator=(const Bash& other) {
+    if (this != &other) {
+        // Copy the queue of ICommand pointers
+        Commands newCommands = other._commands;
+        while (!newCommands.empty()) {
+            ICommand* command = newCommands.front();
+            _commands.emplace(command);
+            newCommands.pop();
+        }
+
+        // Copy the vector of path directories
+        _pathDirectories = other._pathDirectories;
+
+        // Copy the pipe file descriptors
+        _currPipe[0] = other._currPipe[0];
+        _currPipe[1] = other._currPipe[1];
+
+        // Copy the input/output file descriptors
+        _prevPipe[0] = other._prevPipe[0];
+        _prevPipe[1] = other._prevPipe[1];
+    }
+    return *this;
+}
+
+/**
+ * Move assignment operator for the Bash class.
+ * Moves the queue of ICommand pointers, the vector of path directories,
+ * and the pipe file descriptors from the source object to the destination object.
+ * @param other The Bash object to be moved.
+ * @return The reference to the destination object.
+ */
+Bash& Bash::operator=(Bash&& other) noexcept {
+    if (this == &other) {
+        return *this;  // Self-assignment check
+    }
+
+    // Free the resources held by the destination object
+    while (!_commands.empty()) {
+        delete _commands.front();
```

Comment:
If you place in the queue objects that have a destructor, then when `_commands` is destructed it will also destruct gracefully all of its contents. Think about it, is there a justification for using pointers in this queue at all and not just regular instances?